Change KubeVirt VM CPU assignment to not overwrite cpu alloc ratio

soer3n opened the pull request:
**What this PR does / why we need it**:
This pr changes the assignment of cpu resources for kubevirt virtual machines to be able to use kubevirt `cpuAllocationRatio` feature by introducing a new field to the providerSpec called vCPUs.

**Which issue(s) this PR fixes**:
<!--optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged-->
Fixes #1905

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/kind chore
-->

/kind feature

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:
<!--
Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
A new field `cloudProviderSpec.virtualMachine.template.vcpus` was added to kubevirt provider. When `vcpus.cores` is set, the cpus for the virtual machine will be configured in `spec.template.spec.domain.cpu` instead of setting resource requests and limits explicit. The webhook validation will fail if both `vcpus` and `cpus` are configured for a machine.
```

**Documentation**:
<!--
Please do one of the following options:
- Add a link to the existing documentation
- Add a link to the kubermatic/docs pull request
- If no documentation change is applicable then add:
  - TBD (documentation will be added later)
  - NONE (no documentation needed for this PR)
-->
```documentation
NONE
```

ahmedwaleedmalik commented on pkg/cloudprovider/provider/kubevirt/provider.go:
```diff
@@ -840,6 +840,19 @@ func (p *provider) newVirtualMachine(c *Config, pc *providerconfig.Config, machi
 		},
 	}
+
+	if isDedicatedVCPURequested(machine.Spec.Annotations) {
+		cpusAsUint64, err := strconv.ParseUint(c.CPUs, 0, 64)
+
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse cpu cores: %w", err)
+		}
+
+		virtualMachine.Spec.Template.Spec.Domain.CPU = &kubevirtcorev1.CPU{
+			Cores: uint32(cpusAsUint64),
+		}
+	}
```
If this annotation is consumed in machine-controller then this should be configured at `machine.Annotations` not in the spec.

ahmedwaleedmalik commented on sdk/cloudprovider/kubevirt/types.go:
```diff
@@ -25,6 +25,10 @@ import (
 )
 
+const (
+	UseDedicatedKubevirtVCPUAnnotationKey = "kubermatic.io/use-dedicated-kubevirt-cpu"
```
```suggestion
	UseDedicatedKubeVirtVCPUAnnotationKey = "kubermatic.io/use-dedicated-kubevirt-cpu"
```

moadqassem replied:
I think it is better if you actually use the RawConfig instead of annotations.

ahmedwaleedmalik commented on pkg/cloudprovider/provider/kubevirt/provider.go:
```diff
@@ -840,6 +840,19 @@ func (p *provider) newVirtualMachine(c *Config, pc *providerconfig.Config, machi
+
+	if isDedicatedVCPURequested(machine.Spec.Annotations) {
+		cpusAsUint64, err := strconv.ParseUint(c.CPUs, 0, 64)
+
```
No need for space here

ahmedwaleedmalik commented:
1. Please update the release note and mention the annotation explicitly that you have introduced
2. This is a bug or a feature request and not a chore
3. I have updated the title. Please follow proper casing.

soer3n commented:
@ahmedwaleedmalik I changed the kind to feature and modified the release notes accordingly to the changes i added after discussing the implementation with @moadqassem.

ahmedwaleedmalik commented on sdk/cloudprovider/kubevirt/types.go:
```diff
@@ -70,12 +70,20 @@ type Flavor struct {
+// VCPUs.
+type VCPUs struct {
+	Cores   int `json:"cores,omitempty"`
+	Sockets int `json:"sockets,omitempty"`
+	Threads int `json:"threads,omitempty"`
```
Both Sockets and Threads are unused. We should either remove these variables or utilize them at https://github.com/kubermatic/machine-controller/blob/8b52fcc5f41a2d4e5c55b67e51e364e69b50157d/pkg/cloudprovider/provider/kubevirt/provider.go#L893 and https://github.com/kubermatic/machine-controller/blob/8b52fcc5f41a2d4e5c55b67e51e364e69b50157d/pkg/cloudprovider/provider/kubevirt/provider.go#L854-L856

soer3n replied:
I removed the unused values and only left `cores` for now.

ahmedwaleedmalik commented:
Looks much better as part of the providerconfig 🚀  

Same feedback as the first point from https://github.com/kubermatic/machine-controller/pull/1906#issuecomment-2717977729, in release note please explicitly mention the "new field" that you have introduced i.e. `cloudProviderSpec.virtualMachine.template.vcpus`

soer3n commented on sdk/cloudprovider/kubevirt/types.go:
```diff
@@ -70,12 +70,20 @@ type Flavor struct {
 
 // Template.
 type Template struct {
+	VCPUs          VCPUs                          `json:"vcpus,omitempty"`
```
Added explanation to both fields `VCPUs` amd `CPUs`.

soer3n commented on pkg/cloudprovider/provider/kubevirt/provider.go:
```diff
@@ -616,8 +625,12 @@ func (p *provider) Validate(ctx context.Context, _ *zap.SugaredLogger, spec clus
+		}
+
+		if (c.VCPUs == nil && c.Resources.Cpu() == nil) || (c.VCPUs != nil && c.Resources.Cpu() != nil) {
+			return fmt.Errorf("invalid/no cpus configured. Either vpus or cpus have to be configured.")
```
Thanks for the suggestion. I separated both conditions and changed the return message to be a bit more clearer.

ahmedwaleedmalik reviewed (approved):
/approve

soer3n commented:
/cherrypick release/v1.61
